h2o-core/src/main/java/hex/ModelMetricsClustering.java

```diff
@@ -134,35 +129,58 @@ public void reduce(MetricBuilderClustering mm) {
       ArrayUtils.add(_colSumSq, mm._colSumSq);
     }
 
+    /**
+     * Reduce the Metric builder clustering for CV without cluster statistics.
+     *
+     * @param mm metric builder to be reduced
+     */
+    public void reduceForCV(MetricBuilderClustering mm) {
+      super.reduce(mm);
+      ArrayUtils.add(_colSum, mm._colSum);
+      ArrayUtils.add(_colSumSq, mm._colSumSq);
+    }
+
     @Override
     public ModelMetrics makeModelMetrics(Model m, Frame f) {
       assert m instanceof ClusteringModel;
-      ClusteringModel clm = (ClusteringModel) m;
       ModelMetricsClustering mm = new ModelMetricsClustering(m, f, _customMetric);
+      setOverallStatToModelMetrics((ClusteringModel) m, f.numRows(), mm);
+      setCentroidsStatToModelMetrics(mm);
+      return m.addMetrics(mm);
+    }
 
-      mm._size = _size;
+    public ModelMetrics makeModelMetricsForCV(Model m, Frame f) {
```

Comment:
it seems like this method is only needed because the MetricBuilder (mbs[0]) has essentially incorrect information (_size and _within_sumsqe are calculated for one of the models and later other partial information is reduced into it)

perhaps one way to simplify the whole flow and make sure we don't have inconsistent objects is to

define a method on MetricBuilder that would handle the whole aggregation for the per-model metrics, that is create a method that implements this loop:
```    
for (int i = 0; i < N; ++i) {
      if (i > 0) mbs[0].reduceForCV(mbs[i]);
mbs[0]._size = null
mbs[0]._within_sumsqe = null
return mbs[0]
```

then you don't need to differentiate between makeModelMetrics and makeModelMetricsForCV and can just have a one version (that is able to handle the nulls)

Reply:
> it seems like this method is only needed because the MetricBuilder (mbs[0]) has essentially incorrect information (_size and _within_sumsqe are calculated for one of the models and later other partial information is reduced into it)

No, the reason is, that the arrays could have a different size because of `estimate_k=True` parameter. But you are correct, that this method could be merge something like that (because mbs[0]._size and mbs[0[._within_sumsqe are null from initialization):


```
    @Override
    public ModelMetrics makeModelMetrics(Model m, Frame f) {
      assert m instanceof ClusteringModel;
      ClusteringModel clm = (ClusteringModel) m;
      ModelMetricsClustering mm = new ModelMetricsClustering(m, f, _customMetric);
      setOverallStatToModelMetrics((ClusteringModel) m, f.numRows(), mm);
      if(_size != null && _within_sumsqe != null) {
           setCentroidsStatToModelMetrics(mm);
      }
      return m.addMetrics(mm);
    }
```